Documentation/proposals/202212-scrape-config.md

```diff
@@ -0,0 +1,147 @@
+the other `*Monitor` CRDs could be migrated so that they create a ScrapeConfig resource that would ultimately be used by
+the operator to generate scrape configuration.
+
+## Pitfalls of the current solution
+
+Using `additionalScrapeConfig` comes with drawbacks:
+
+* Teams have to build an infrastructure to add scrape rules in a centralized manner, which creates a bottleneck since a
+  single team becomes responsible for the configuration
+* There is no input validation, which can lead to an invalid prometheus configuration
+
+# Goals
+
+* Provide a way for users to self-service adding scrape targets
+* Consolidate the scrape configuration generation logic in a central point for other resources to use
+
+## Audience
+
+* Users who serve Prometheus as a service and want to have their customers autonomous in defining scrape configs
+* Users who want to manage scrape configs the same way as for services running within the Kubernetes cluster
+* Users who want a supported Kubernetes way of scraping targets outside the Kubernetes cluster
+
+# Non-Goals
+
+* This proposal doesn't aim at covering all the fields in
+  [`<scrape_config>`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#scrape_config).
+  Specifically, no service discovery other than `static_configs` or `file_sd_configs` should be implemented at first.
+* Refactoring of the other CRDs is not in scope for the first version
+
+# How
+
+As described by
+[@aulig in #2787](https://github.com/prometheus-operator/prometheus-operator/issues/2787#issuecomment-559776221), we
+will create a new ScrapeConfig CRD, this new CRD will act the same as the other CRDs and append scrape configurations to
+the configuration. Usage of ScrapeConfig doesn't exclude the use of the other CRDs, they are not mutually exclusive.
+`ScrapeConfig` will allow for any scraping configuration, while the other CRDs provide sane defaults. This will allow
+for isolated testing of the new `ScrapeConfig` CRD.
+
+```mermaid
+graph TD;
+ServiceMonitor --> prometheusConfig
+PodMonitor --> prometheusConfig
+ScrapeConfig --> prometheusConfig
+prometheusConfig
+```
+
+Using a pseudo custom resource definition, we should have the following:
+
+```yaml
+apiVersion: monitoring.coreos.com/v1alpha1
+kind: ScrapeConfig
+metadata:
+  name: my-scrape-config
+  namespace: system-monitoring
+  labels:
+    test: value
+spec:
+  staticConfigs:
+    - <StaticConfig>[] # new resource
+  fileSDConfigs:
+    - <fileSDConfig>[] # new resource
+  httpSDConfigs:
+    - <httpSDConfig>[] # new resource
+  relabelConfigs:
+    - <RelabelConfig>[] # https://github.com/prometheus-operator/prometheus-operator/blob/e4e27052f57040f073c6c1e4aedaecaaec77d170/pkg/apis/monitoring/v1/types.go#L1150
+  metricsPath: /metrics
+```
+
+with the following new resources:
+
+* `staticConfig`
+
+```yaml
+targets:
+  - target:9100
+labels:
+  labelA: placeholder
+```
+
+* `fileSDConfig`:
```

Comment:
After giving it more thoughts, I'm wondering if file SD configuration is really needed and if it brings enough value compared to static targets? The user experience with file SD would be:
* create the configmaps/secrets for targets
* create the scrape config object
* update the configmaps/secrets
With the static config, the user has to deal only with the scrape config object which feels less burden and more idiomatic. There's also the challenge of syncing the user's secrets/configmaps to an operator-owned secret as well as avoiding the redeployment of Prometheus on secret addition/removal (similar to #5085). Finally it's less obvious how the operator reports that the user's secret/configmap is invalid.

From a classic Prometheus user's standpoint, file SD is interesting because compared to static config, Prometheus can pick up changes almost immediately. But in the operator's case, it's more an implementation detail and it might be possible to "translate" static targets into a file SD configuration?

Reply:
I was also wondering about this. @xiu afair you mentioned that you have a use case for file sd with the operator, could you share more?

Reply:
Sure thing. I added file_sd as I'm mostly interested in this one. Our usecase is that we have a job pulling discovery data from internal systems and landing it in a configmap regularly like:
```json
[
  {
    "labels": {
      "rack": "1",
      "cluster": "clusterA",
      "type": "typeA",
      "team": "teamB"
    },
    "targets": [
      "host1.fqdn"
    ]
  },
  {
    "labels": {
      "rack": "2",
      "cluster": "clusterB",
      "type": "typeB",
      "team": "teamB"
    },
    "targets": [
      "host2.fqdn"
    ]
  }
]
```

With this configmap mounted in Prometheus, we can then create scraping rules like this, that we use `additionalScrapeConfigs` for:
```yaml
- file_sd_configs:
  - files:
    - ../configmaps/inventory/file.json
  job_name: my_job
  relabel_configs:
    - action: keep
      regex: typeA
      source_labels:
      - type
    - regex: (.*)
      replacement: ${1}:9100
      source_labels:
      - __address__
      target_label: __address__
  static_configs: []
```

The UX is not great of course, and we'll be working at revamping that, but it is the system we have today.

In general though, I would imagine `ScrapeConfig` to stick with Prometheus' primitives and have other higher level CRDs (`ServiceMonitor`/`PodMonitor`) be more opinionated. I'm happy to jump on a call tomorrow to describe my usecase in more details if needed.

Reply:
I think that's a valid use-case and others might have a similar one. However, it somehow feels like rebranded `additionalScrapeConfigs` :thinking: Do we want to also have validation on files listed in a ConfigMap?

There is also slight implementation detail we need to remember in this particular case - since ConfigMaps/Secrets are double linked when mounting in a container, we cannot rely on Prometheus just picking up newly changed files from those ConfigMap/Secrets. This means either config-reloader sidecar would need to be updated every time a new `file_sd_config` is added (which results in prometheus redeploy) or the operator would need to somehow track changes in those ConfigMaps (annotation with a checksum?) or some other way of ensuring this. All this means we are adding quite a ton of complexity to the initial implementation and without this particular feature I don't think file_sd would be complete.

Reply:
> However, it somehow feels like rebranded additionalScrapeConfigs 🤔
It's one of the goals imo, provide Kubernetes primitives rather than a centralized `additionalScrapeConfigs`.

> Do we want to also have validation on files listed in a ConfigMap?
I think checking that the file exist would already be good. I wouldn't consider validation of the content as needed, at least in the first version. Worst case, we won't discover scrape targets.

> There is also slight implementation detail we need to remember in this particular case - since ConfigMaps/Secrets are double linked when mounting in a container, we cannot rely on Prometheus just picking up newly changed files from those ConfigMap/Secrets.
I'm not sure I get this one though I'm far from the most knowledgeable on the code base 😄 Currently we use one of these configmaps without any issue and it gets updated regularly. Prometheus has [`refresh_interval` on `file_sd_configs`](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#file_sd_config) set to 5m by default to reload the file. Am I missing something?

Reply:
All good and thank you for the explanation. I may have overthinked it :)

Reply:
> I think checking that the file exist would already be good. I wouldn't consider validation of the content as needed, at least in the first version. Worst case, we won't discover scrape targets.

If someone crafts an invalid file SD config and the operator passes it as-is to Prometheus, it's still a sub-par experience IMHO.

The area I'm not clear about is how the operator will synthesize the file SD configs from all the ScrapeConfigs and how they will be passed down to the statefulset. I'll imagine that the operator will copy the secret/configmap data to a common secret so it avoids redeploying the pods everytime a ScrapeConfig is added/removed/updated? We also have to deal deal with the secret's limit size (1MiB). So I'd like to see a bit more details in the proposal :)

Reply:
> If someone crafts an invalid file SD config and the operator passes it as-is to Prometheus, it's still a sub-par experience IMHO.

Absolutely, this can come as we iterate. I wouldn't put this as part of the first version, the benefits of the CRD is still very high compared to the current solution.

> The area I'm not clear about is how the operator will synthesize the file SD configs from all the ScrapeConfigs and how they will be passed down to the statefulset. I'll imagine that the operator will copy the secret/configmap data to a common secret so it avoids redeploying the pods everytime a ScrapeConfig is added/removed/updated? We also have to deal deal with the secret's limit size (1MiB). So I'd like to see a bit more details in the proposal :)

The same way we do for `ServiceMonitor`, each `ScrapeConfig` will be passed down to the configuration. The `ScrapeConfig` will not manipulate a configmap nor a secret, the filepath will be a string and it'll be the responsibility of the operator to mount the file via `Prometheus.ConfigMaps`. The operator is not responsible for the SD file.

Reply:
I realize the confusion now as I referenced `SecretOrConfigMap` in the proposal:

```yaml
files:
  - <SecretOrConfigMap>[] # https://github.com/prometheus-operator/prometheus-operator/blob/e4e27052f57040f073c6c1e4aedaecaaec77d170/pkg/apis/monitoring/v1/types.go#L1644
refresh_interval: 5m
```

Updating the document to reflect what I just said.

Reply:
Done in https://github.com/prometheus-operator/prometheus-operator/pull/5279/commits/9f1cd504392ffd1cda5f05f0c5fe1b64f99dbc3a, @simonpasquier @paulfantom sorry for the confusion there and thanks for the thorough review, let me know what you think with this change :)